src/pykeen/losses.py

```diff
@@ -953,6 +954,68 @@ def forward(
         return loss
 
 
+@parse_docdata
+class FocalLoss(PointwiseLoss):
+    """
+    Implementation of the focal loss as a module.
+
+    Inspired by the functional form from torchvision.
+
+    Proposed by [lin2018]_.
+
+    # TODO: Move to correct place
+    Focal Loss for Dense Object Detection
+    Tsung-Yi Lin, Priya Goyal, Ross Girshick, Kaiming He, Piotr Dollár
+    ICCV'17.
+    https://arxiv.org/abs/1708.02002
+
+    .. seealso ::
+        https://pytorch.org/vision/stable/_modules/torchvision/ops/focal_loss.html
+
+    ---
+    name: Focal
+    """
+
+    def __init__(
+        self,
+        *,
+        gamma: float = 2.0,
+        alpha: Optional[float] = None,
+        **kwargs,
+    ):
+        """
+        Initialize the loss module.
+
+        :param gamma:
+            Exponent of the modulating factor (1 - p_t) to balance easy vs hard examples.
+        :param alpha:
+            Weighting factor in range (0,1) to balance positive vs negative examples.
+        :param kwargs:
+            Additional keyword-based arguments passed to PointwiseLoss.
+        """
+        super().__init__(**kwargs)
+        if alpha is not None and not (0 < alpha < 1):
+            raise ValueError(f"If alpha is provided, it must be from (0, 1), i.e. the open interval, but it is {alpha}")
+        self.alpha = alpha
+        self.gamma = gamma
+
+    def forward(
+        self,
+        prediction: torch.FloatTensor,
+        labels: torch.FloatTensor,
+    ) -> torch.FloatTensor:  # noqa: D102
+        p = prediction.sigmoid()
+        ce_loss = functional.binary_cross_entropy_with_logits(prediction, labels, reduction="none")
+        p_t = p * labels + (1 - p) * (1 - labels)
```

Comment:
is there a more general way to define losses that do this `p * labels + (1 - p) * (1 - labels)` calculation?

Reply:
focal loss is an extension of binary_cross_entropy, but I did not merge them into a single class, since there is a torch builtin for computing BCE, which seems to [do some magic](https://pytorch.org/docs/stable/_modules/torch/nn/functional.html#binary_cross_entropy_with_logits).

Both are special cases of [cross-entropy](https://en.wikipedia.org/wiki/Cross_entropy)-based losses, seem from the `p log q` part (where `p` is labels, and `log q` is predictions :sweat_smile: ).

They are for the 2-class setting, where it is sufficient to only have label/logits for one class, since the other can be derived from it (that is the `1-p` / `1 - labels` part).

**TL;DR:** I do not see a useful way to generalize this now (from the code side), since up until now we only have two of them, and one makes uses of a specialized implementation outside of our lib.